### PR TITLE
Feature/post type enhancements

### PR DESCRIPTION
**Pull request by mustafauysal**
## Description of the Change

This PR tries to address two problems, related to `post_type` support. 

* In order to restrict the autosuggest query by post_type. I have added a control that checks `[name="post_type"]` input and applying particular `post_type` to query. Therefore, we can show only the "products" for WooCommerce's own search bar. 

* I have changed facet's query build logic a bit, now `ep_facet_allowed_query_args` allows to extend allowed query parameters. `post_type` and `s` are supported by default. 

### Alternate Designs

* Perhaps, we can add a new flag item for Autosuggest's localize script to understand that request made from a product page

### Benefits

Improving `post_type` support without adding custom code

### Possible Drawbacks

* I don't consider as drawbacks but we likely need to re-apply autosuggest related changes to #1663

### Verification Process

[Autosuggest]
* Install WooCommerce
* Activate EP's Woo and Autosuggest feature
* Visit `/shop/` page
* Enter the search query and see the suggested results 

[Facets]
* Make a WooCommerce product search
* Check facet's widgets have `post_type` on the target url

### Screenshot
(first input Woo's search, second is generic) 

Screen Shot 2020-03-05 at 15 49 25

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

### Applicable Issues

#1687, #1684

### Changelog Entry

* Autosuggest: respects the `[name="post_type"]` input in the same form.
* Additional query parameters can register through `ep_facet_allowed_query_args` for facets

### CODE REVIEW

**Comment by JakePT** [comment]
I have update the PR so that the JavaScript changes to autosuggest have been applied to an up to date version of that script, and don't use jQuery. 

I also updated the JavaScript to support different possible post type inputs in the search form, including:

- a single hidden input, as used by WooComerce.
- multiple hidden post type inputs that use `post_type[]` to query multiple post types.
- post type checkboxes using `post_type[]`.
- post type radio buttons that use `post_type`.
- post type radio buttons that use `post_type`, with a hidden `post_type` input as a default fallback.

All of those are possible with searchform.php, so this version ensures that the autosuggest behaviour matches the behaviour of submitting the form.

I used a couple `FormData` methods for this which aren't supported in IE, like `getAll()`, and I'm not certain if the build process for this plugin will handle IE11 support for that, so that will need to be tested.

This probably also addresses #2135.

**Comment by oscarssanchez** [comment]
@JakePT can you let me know how the post type feature is planned to work? I was thinking it should add a post type support in the facet widget but seems it is not the case.

**Comment by JakePT** [comment]
> @JakePT can you let me know how the post type feature is planned to work? I was thinking it should add a post type support in the facet widget but seems it is not the case.

It just limits autosuggest results to whichever post type is being searched by the search form. The main example is WooCommerce, which uses a hidden `post_type` field in the form to search products. This picks up on that so that autosuggest only shows products.

**Comment by felipeelia** [comment]
Let's please adjust the `@since` tag before merging

File: assets/js/autosuggest.js [hunk 1]
```diff
@@ -565,7 +586,21 @@ function init() {
 		if (searchText.length >= 2) {
 			setFormIsLoading(true, input);
 
-			const query = buildSearchQuery(searchText, placeholder, queryJSON);
+			let query = buildSearchQuery(searchText, placeholder, queryJSON);
+
+			if (postTypes.length > 0) {
+				query = JSON.parse(query);
+
+				if (typeof query.post_filter.bool.must !== 'undefined') {
```
[the remark this answers is not in the document]
Comment (JakePT) [reply assets/js/autosuggest.js:594]: @oscarssanchez Yeah, the post type could be different based on the input being used.

**Comment by JakePT** [comment]
> Let's please adjust the `@since` tag before merging @JakePT and @oscarssanchez

Done @felipeelia

**Review by JakePT — APPROVED** [review]

**Review by Rahmon — APPROVED** [review]

**Pull request merged**